File: android/src/main/kotlin/dev/brewkits/native_workmanager/NativeWorkmanagerPlugin+Middleware.kt
package dev.brewkits.native_workmanager

import android.content.Context
import dev.brewkits.native_workmanager.store.MiddlewareStore
import dev.brewkits.native_workmanager.utils.MappingUtils.toJson
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

internal fun NativeWorkmanagerPlugin.handleRegisterMiddleware(call: MethodCall, result: Result) {
    scope.launch {
        try {
            val type = call.argument<String>("type")
                ?: return@launch result.error("INVALID_ARGS", "type required", null)
            val configMap = call.arguments as Map<String, Any?>
            val configJson = toJson(configMap)

            NativeLogger.d("🛡️ Registering middleware: $type")

            withContext(Dispatchers.IO) {
                middlewareStore.add(type, configJson)
            }

            result.success(null)
        } catch (e: Exception) {
            NativeLogger.e("❌ Register middleware error", e)
            result.error("MIDDLEWARE_ERROR", e.message, null)
        }
    }
}

/**
 * Applies registered middleware to a worker configuration.
 *
 * This can be called from any worker before execution.
 * Internal top-level function so it can be imported from utils sub-packages.
 */
internal fun applyMiddlewareInternal(context: Context, workerClassName: String, configJson: String): String {
    try {
        val store = MiddlewareStore.getInstance(context)
        val middlewares = store.getAll()
        if (middlewares.isEmpty()) return configJson

        val config = JSONObject(configJson)
        var modified = false

        for (mw in middlewares) {
            val mwConfig = JSONObject(mw.configJson)
            when (mw.type) {
                "header" -> {
                    if (applyHeaderMiddleware(config, mwConfig)) {
                        modified = true
                    }
                }
                "remoteConfig" -> {
                    if (applyRemoteConfigMiddleware(config, mwConfig, workerClassName)) {
                        modified = true
                    }
                }
            }
        }

        return if (modified) config.toString() else configJson
    } catch (e: Exception) {
        NativeLogger.e("❌ Error applying middleware", e)
        return configJson
    }
}

/** Companion wrapper for callers that use [NativeWorkmanagerPlugin.applyMiddleware]. */
fun NativeWorkmanagerPlugin.Companion.applyMiddleware(
    context: Context,
    workerClassName: String,
    configJson: String,
): String = applyMiddlewareInternal(context, workerClassName, configJson)

private fun applyRemoteConfigMiddleware(
    workerConfig: JSONObject,
    mwConfig: JSONObject,
    workerClassName: String
): Boolean {
    val targetType = mwConfig.optString("workerType").takeIf { it.isNotEmpty() }
    if (targetType != null && !workerClassName.contains(targetType, ignoreCase = true)) {
        return false
    }
    val values = mwConfig.optJSONObject("values") ?: return false
    var modified = false
    val keys = values.keys()
    while (keys.hasNext()) {
        val key = keys.next()
        workerConfig.put(key, values.get(key))
        modified = true
    }
    return modified
}

private fun applyHeaderMiddleware(workerConfig: JSONObject, mwConfig: JSONObject): Boolean {
    val url = workerConfig.optString("url", null) ?: return false
    val pattern = mwConfig.optString("urlPattern", null)
    
    // Check if URL matches pattern (if provided)
    if (pattern != null && !Regex(pattern).containsMatchIn(url)) {
        return false
    }

    val headersToAdd = mwConfig.optJSONObject("headers") ?: return false
    val workerHeaders = workerConfig.optJSONObject("headers") ?: JSONObject().also { workerConfig.put("headers", it) }

    val keys = headersToAdd.keys()
    while (keys.hasNext()) {
        val key = keys.next()
        workerHeaders.put(key, headersToAdd.get(key))
    }

    return true
}

/**
 * Fire-and-forget HTTP POST for LoggingMiddleware.
 *
 * Called after each task completes (success or failure). Finds all registered
 * LoggingMiddleware records and POSTs task execution metadata to each logUrl.
 * Errors are logged but never propagated — logging must never affect worker results.
 */
internal fun NativeWorkmanagerPlugin.applyLoggingMiddleware(
    taskId: String,
    workerClassName: String,
    success: Boolean,
    message: String?,
    durationMs: Long?,
    workerConfig: String?
) {
    try {
        val middlewares = middlewareStore.getAll().filter { it.type == "logging" }
        if (middlewares.isEmpty()) return

        for (mw in middlewares) {
            val mwConfig = try { JSONObject(mw.configJson) } catch (_: Exception) { continue }
            val logUrl = mwConfig.optString("logUrl").takeIf { it.isNotEmpty() } ?: continue
            val includeConfig = mwConfig.optBoolean("includeConfig", false)

            scope.launch(Dispatchers.IO) {
                try {
                    val payload = JSONObject().apply {
                        put("taskId", taskId)
                        put("workerClassName", workerClassName)
                        put("success", success)
                        put("timestamp", System.currentTimeMillis())
                        if (durationMs != null) put("durationMs", durationMs)
                        if (!message.isNullOrEmpty()) put("message", message)
                        if (includeConfig && workerConfig != null) {
                            try { put("workerConfig", JSONObject(workerConfig)) } catch (_: Exception) {}
                        }
                    }
                    val conn = URL(logUrl).openConnection() as HttpURLConnection
                    try {
                        conn.requestMethod = "POST"
                        conn.setRequestProperty("Content-Type", "application/json; charset=utf-8")
                        conn.doOutput = true
                        conn.connectTimeout = 5_000
                        conn.readTimeout = 5_000
                        conn.outputStream.use { it.write(payload.toString().toByteArray(Charsets.UTF_8)) }
                        conn.responseCode // trigger the request
                    } finally {
                        conn.disconnect()
                    }
                } catch (e: Exception) {
                    NativeLogger.e("LoggingMiddleware: Failed to POST to $logUrl for task '$taskId'", e)
                }
            }
        }
    } catch (e: Exception) {
        NativeLogger.e("LoggingMiddleware: Unexpected error for task '$taskId'", e)
    }
}

File: ios/native_workmanager/Sources/native_workmanager/NativeWorkmanagerPlugin+Middleware.swift
import Foundation
import Flutter

extension NativeWorkmanagerPlugin {

    internal func handleRegisterMiddleware(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let type = args["type"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "type required", details: nil))
            return
        }

        guard let configData = try? JSONSerialization.data(withJSONObject: args),
              let configJson = String(data: configData, encoding: .utf8) else {
            result(FlutterError(code: "SERIALIZATION_ERROR", message: "Failed to serialize middleware config", details: nil))
            return
        }

        if #available(iOS 13.0, *) {
            MiddlewareStore.shared.add(type: type, configJson: configJson)
        }

        NativeLogger.d("🛡️ Registering middleware: \(type)")
        result(nil)
    }

    /// Applies registered middleware to a worker configuration.
    public static func applyMiddleware(workerClassName: String, config: [String: Any]) -> [String: Any] {
        if #available(iOS 13.0, *) {
            let middlewares = MiddlewareStore.shared.getAll()
            if middlewares.isEmpty { return config }

            var resultConfig = config
            var modified = false

            for mw in middlewares {
                guard let mwData = mw.configJson.data(using: .utf8),
                      let mwConfig = try? JSONSerialization.jsonObject(with: mwData) as? [String: Any] else {
                    continue
                }

                switch mw.type {
                case "header":
                    if applyHeaderMiddleware(workerConfig: &resultConfig, mwConfig: mwConfig) {
                        modified = true
                    }
                case "remoteConfig":
                    if applyRemoteConfigMiddleware(workerConfig: &resultConfig, mwConfig: mwConfig, workerClassName: workerClassName) {
                        modified = true
                    }
                case "logging":
                    // LoggingMiddleware fires post-execution via applyLoggingMiddleware().
                    // It does not modify worker config — skip here.
                    break
                default:
                    break
                }
            }

            return modified ? resultConfig : config
        }
        return config
    }

    private static func applyRemoteConfigMiddleware(
        workerConfig: inout [String: Any],
        mwConfig: [String: Any],
        workerClassName: String
    ) -> Bool {
        if let targetType = mwConfig["workerType"] as? String, !targetType.isEmpty {
            guard workerClassName.range(of: targetType, options: .caseInsensitive) != nil else {
                return false
            }
        }
        guard let values = mwConfig["values"] as? [String: Any], !values.isEmpty else {
            return false
        }
        for (key, value) in values {
            workerConfig[key] = value
        }
        return true
    }

    /// Fire-and-forget HTTP POST for LoggingMiddleware.
    ///
    /// Called after each task completes (success or failure). Finds all registered
    /// LoggingMiddleware records and POSTs task execution metadata to each logUrl.
    /// Errors are logged but never propagated — logging must never affect worker results.
    public static func applyLoggingMiddleware(
        taskId: String,
        workerClassName: String,
        success: Bool,
        message: String?,
        durationMs: Int64?
    ) {
        guard #available(iOS 13.0, *) else { return }
        let middlewares = MiddlewareStore.shared.getAll().filter { $0.type == "logging" }
        guard !middlewares.isEmpty else { return }

        Task.detached(priority: .utility) {
            for mw in middlewares {
                guard let mwData = mw.configJson.data(using: .utf8),
                      let mwConfig = try? JSONSerialization.jsonObject(with: mwData) as? [String: Any],
                      let logUrl = mwConfig["logUrl"] as? String, !logUrl.isEmpty,
                      let url = URL(string: logUrl) else { continue }

                var payload: [String: Any] = [
                    "taskId": taskId,
                    "workerClassName": workerClassName,
                    "success": success,
                    "timestamp": Int64(Date().timeIntervalSince1970 * 1000)
                ]
                if let d = durationMs { payload["durationMs"] = d }
                if let m = message, !m.isEmpty { payload["message"] = m }

                guard let body = try? JSONSerialization.data(withJSONObject: payload) else { continue }

                var request = URLRequest(url: url, timeoutInterval: 5)
                request.httpMethod = "POST"
                request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
                request.httpBody = body

                do {
                    _ = try await URLSession.shared.data(for: request)
                } catch {
                    print("LoggingMiddleware: Failed to POST to \(logUrl) for task '\(taskId)': \(error)")
                }
            }
        }
    }

    private static func applyHeaderMiddleware(workerConfig: inout [String: Any], mwConfig: [String: Any]) -> Bool {
        guard let url = workerConfig["url"] as? String else { return false }

        if let pattern = mwConfig["urlPattern"] as? String {
            let regex = try? NSRegularExpression(pattern: pattern)
            let range = NSRange(location: 0, length: url.utf16.count)
            if regex?.firstMatch(in: url, options: [], range: range) == nil {
                return false
            }
        }

        // Accept [String: Any] so numeric/boolean header values are not silently dropped
        guard let headersToAdd = mwConfig["headers"] as? [String: Any] else { return false }
        var workerHeaders = workerConfig["headers"] as? [String: Any] ?? [:]

        for (key, value) in headersToAdd {
            workerHeaders[key] = value
        }

        workerConfig["headers"] = workerHeaders
        return true
    }
}
